File: src/pkg/player/8durt/HeadTerminal.lua
-- package: programs

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "HeadTerminal.kt", {["1-12"]=1,["13"]=36,["14"]=37,["15"]=38,["16"]=39,["17-18"]=40,["19"]=43,["20"]=44,["21"]=45,["22-23"]=46,["24"]=49,["25"]=50,["26-30"]=51,["31"]=57,["32"]=58,["33-37"]=59,["38"]=64,["39"]=69,["40"]=70,["41"]=71,["42-45"]=72,["46"]=80,["47"]=81,["48-49"]=82,["50"]=84,["51"]=85,["52"]=86,["53-54"]=87,["55"]=89,["56"]=90,["57-63"]=91}, "programs")
ktox_require("lib/RoleCheck")
ktox_require("lib/DepositChest")
ktox_require("lib/Farm")
ktox_require("lib/Cli")
ktox_require("lib/Dashboard")
ktox_require("lib/PassiveFeeder")

local function main()
    println("Head terminal starting...")
    local hasModem = ktoxRednetOpenAny()
    if not hasModem then
        println("No modem found - attach one and reboot. A head terminal needs one to detect other heads and talk to secondaries.")
        return
    end
    local existingHead = queryForHead(2.0)
    if existingHead ~= -1 then
        println("Another head is already running (id " .. tostring(existingHead) .. ") - refusing to start. Only one head terminal is allowed on the network.")
        return
    end
    println("Head terminal ready. Use the touch dashboard (monitor if attached, otherwise this screen).")
    while true do
        parallel.waitForAny(function()
            return handleLocalInput()
        end, function()
            return handleRemoteMessage()
        end)
        topUpPassiveFeeders()
        manageFarms()
        drainDepositChests()
    end
end

function handleLocalInput()
    local commandLine = runDashboardLoop()
    showDashboardBusy("Working...")
    local result = runCliCommand(commandLine)
    println(result)
    showDashboardResult(result)
end

function handleRemoteMessage()
    local got = ktoxRednetReceiveAny(3600.0)
    if not got then
        return
    end
    local senderId = ktoxRednetLastSenderId()
    local protocol = ktoxRednetLastProtocol()
    if protocol == VAULT_ROLE_QUERY_PROTOCOL then
        rednet.send(senderId, "head", VAULT_ROLE_REPLY_PROTOCOL)
    elseif protocol == VAULT_CMD_PROTOCOL then
        local commandLine = ktoxRednetLastMessage()
        local result = runCliCommand(commandLine)
        rednet.send(senderId, result, VAULT_RESULT_PROTOCOL)
    end
end


-- Auto-generated call to main function
main()

File: ktox-lua-cc/src/main/kotlin/programs/HeadTerminal.kt
package programs

import common.ktoxRednetLastMessage
import common.ktoxRednetLastProtocol
import common.ktoxRednetLastSenderId
import common.ktoxRednetReceiveAny
import common.parallelWaitForAny
import common.readInput
import common.rednetOpenAny
import common.rednetSend
import common.termWrite
import lib.VAULT_CMD_PROTOCOL
import lib.VAULT_ROLE_QUERY_PROTOCOL
import lib.VAULT_ROLE_REPLY_PROTOCOL
import lib.VAULT_RESULT_PROTOCOL
import lib.manageFarms
import lib.queryForHead
import lib.runCliCommand
import lib.topUpPassiveFeeders

// The head vault terminal (see PLAN.md "Terminal roles") — the sole
// decision-maker. Runs the shared CLI dispatcher (lib/Cli.kt) against
// both its own local read() prompt and commands forwarded over rednet
// from secondary terminals, multiplexed via parallel.waitForAny (see
// common/Parallel.kt — UNVERIFIED, this whole rednet+parallel path has
// not been exercised in a real game; validate with two real computers
// before relying on it).
//
// Usage: HeadTerminal (no args)

fun main() {
    println("Head terminal starting...")
    val hasModem = rednetOpenAny()
    if (!hasModem) {
        println("No modem found - attach one and reboot. A head terminal needs one to detect other heads and talk to secondaries.")
        return
    }

    val existingHead = queryForHead(2.0)
    if (existingHead != -1) {
        println("Another head is already running (id ${existingHead}) - refusing to start. Only one head terminal is allowed on the network.")
        return
    }

    println("Head terminal ready. Commands: list, pull, craft, trash.")
    while (true) {
        parallelWaitForAny(
            { handleLocalInput() },
            { handleRemoteMessage() },
        )
        // Opportunistic, not on an independent timer — see
        // lib/PassiveFeeder.kt for why.
        topUpPassiveFeeders()
        manageFarms()
    }
}

fun handleLocalInput() {
    termWrite("> ")
    val commandLine = readInput()
    println(runCliCommand(commandLine))
}

fun handleRemoteMessage() {
    // Long finite timeout rather than "forever" - no evidence either way
    // on whether ktox correctly transpiles a sentinel for "no timeout"
    // through this native binding, so a large number sidesteps the
    // question entirely (worst case: re-loops once an hour for nothing).
    val got = ktoxRednetReceiveAny(3600.0)
    if (!got) {
        return
    }
    val senderId = ktoxRednetLastSenderId()
    val protocol = ktoxRednetLastProtocol()
    if (protocol == VAULT_ROLE_QUERY_PROTOCOL) {
        rednetSend(senderId, "head", VAULT_ROLE_REPLY_PROTOCOL)
    } else if (protocol == VAULT_CMD_PROTOCOL) {
        val commandLine = ktoxRednetLastMessage()
        val result = runCliCommand(commandLine)
        rednetSend(senderId, result, VAULT_RESULT_PROTOCOL)
    }
}
